design/001-secretsink.md

```diff
@@ -0,0 +1,188 @@
+      namespace: "a-team"
+      caBundle: "..."
+      caProvider:
+        type: "Secret"
+        name: "my-cert-secret"
+        key: "cert-key"
+      auth:
+        tokenSecretRef:
+          name: "my-secret"
+          namespace: "secret-admin"
+          key: "vault-token"
+        appRole:
+          path: "approle"
+          roleId: "db02de05-fa39-4855-059b-67221c5c2f63"
+          secretRef:
+            name: "my-secret"
+            namespace: "secret-admin"
+            key: "vault-token"
+        kubernetes:
+          mountPath: "kubernetes"
+          role: "demo"
+          serviceAccountRef:
+            name: "my-sa"
+            namespace: "secret-admin"
+          secretRef:
+            name: "my-secret"
+            namespace: "secret-admin"
+            key: "vault"
+    gcpsm:
+      encryptionConfig: {} # Specific config for Creating Secrets GCP SM
+      auth:
+        secretRef:
+          secretAccessKeySecretRef:
+            name: gcpsm-secret
+            key: secret-access-credentials
+      projectID: myproject
+```
+
+```yaml
+apiVersion: external-secrets.io/v1alpha1
+kind: SecretSink
+metadata:
+  name: "hello-world"
+  namespace: my-ns # Same of the SecretStores
+spec:
+  secretStoreRefs:
+  - name: secret-store
+    kind: SecretStore
+  - name: secret-store-2
+    kind: SecretStore
+  - name: cluster-secret-store
+    kind: ClusterSecretStore
+  refreshInterval: "1h"
+
+  data:
+    selector:
+      secret:
+        name: foobar
+    match:
+    - secretKey: foobar
+      remoteRefs:
+      - remoteKey: my/path/foobar 
+        property: my-property #optional. To allow coming back from a 'dataFrom'
+      - remoteKey: secret/my-path-foobar
+        property: another-property
+    rewrite:
+    - secretKey: game-(.+).(.+)
+      remoteRefs:
+      - remoteKey: my/path/($1) 
+        property: prop-($2)
+        secretStores:
+        - name: secret-store #Applies this way only to 'secret-store' secretStore
+      - remoteKey: my-path-($1)-($2) #Applies this way to all other secretStores
+
+status:
+  refreshTime: "2019-08-12T12:33:02Z"
+  conditions:
+  - type: Ready
+    status: "True" 
+    reason: "SecretSynced"
+    message: "Secret was synced" #Fully synced
+    lastTransitionTime: "2019-08-12T12:33:02Z"
+  - type: Ready
+    status: "True"
+    reason: "SecretSyncError"
+    message: "Secret sync failed to Sink SecretStore: abc"
+    lastTransitionTime: "2019-08-12T12:33:02Z"
+  - type: Ready
+    status: "False"
+    reason: "SecretSyncError"
+    message: "Secret sync failed to Sink SecretStore: abc, def"
+    lastTransitionTime: "2019-08-12T12:33:02Z"
+```
+
+### Behavior
+When checking SecretSink for the Source Secret, check existing labels for SecretStore reference of that particular Secret. If this SecretStore reference is an object in SecretSink SecretStore lists, a SecretSyncError should be emited as we cannot sink the secret to the same SecretStore.
```

Comment:
```suggestion
When checking SecretSink for the Source Secret, check existing labels for SecretStore reference of that particular Secret. If this SecretStore reference is an object in SecretSink SecretStore lists, a SecretSyncError should be emited as we cannot sync the secret to the same SecretStore.
```

I would give the user a urgent warning (in the docs) that this is NOT something that someone should do. We can not guarantee to catch user-created sync loops. It is quite likely that a user creates two stores: one for reading and one for writing. We can not catch this particular case.

Reply:
I agree with you. We should make it clear in the docs it's not something they can do. I am concerned with honest configuration errors (like forgetting to change a region or something like that), so I was trying to figure _some_ sort of protection at software level.